### PR TITLE
Rebased aggregate

### PR DESCRIPTION
**Pull request by NivLipetz**

### CODE REVIEW

File: src/reports/models/reportsManager.js [hunk 1]
```diff
@@ -102,4 +130,72 @@ function generateGraphanaUrl(report, grafanaUrl) {
...
+    if (allSubscribersFinished(uniqueSubscribersStages)) {
+        return constants.REPORT_FINISHED_STATUS;
+    } else if (Date.now() >= reportStartTimeMs + reportDurationMs + delayedTimeInMs) {
+        if (uniqueSubscribersStages.includes(constants.SUBSCRIBER_DONE_STAGE)) {
+            return constants.REPORT_PARTIALLY_FINISHED_STATUS;
+        } else {
+            return constants.REPORT_FAILED_STATUS;
+        }
+    } else if (allSubscribersInitializing(uniqueSubscribersStages)) {
+        return constants.REPORT_INITIALIZING_STATUS;
```
Comment (enudler) [inline src/reports/models/reportsManager.js]: all this code can be little more simple
if (unique.length === 1) {
return statusMap[unique[0]];
}

File: src/reports/models/reportsManager.js [hunk 2]
```diff
@@ -102,4 +130,72 @@ function generateGraphanaUrl(report, grafanaUrl) {
...
+        return constants.REPORT_FAILED_STATUS;
+    }
+}
+
+function getListOfSubscribersStages (report) {
+    const runnerStates = report.subscribers.map((subscriber) => subscriber.stage);
+    return runnerStates;
+}
+
+function allSubscribersFinished (subscriberStages) {
```
Comment (enudler) [inline src/reports/models/reportsManager.js]: same function different param 
can be refactored

File: src/reports/models/database/cassandra/cassandraConnector.js [hunk 3]
```diff
@@ -75,6 +81,30 @@ function getStats(testId, reportId) {
     return executeQuery(GET_REPORT_STATS, params, queryOptions);
 }
 
+function subscribeRunner(testId, reportId, runnerId) {
+    let params;
+    params = [testId, reportId, runnerId, constants.SUBSCRIBER_INITIALIZING_STAGE];
```
Comment (enudler) [inline src/reports/models/database/cassandra/cassandraConnector.js]: dont forget to move the const to the manager

File: src/reports/models/database/cassandra/cassandraConnector.js [hunk 4]
```diff
@@ -88,3 +118,22 @@ function executeQuery(query, params, queryOptions) {
         return Promise.reject(new Error('Error occurred in communication with cassandra'));
     });
 }
+
+async function getReportsAndParse(query, params, queryOptions) {
```
Comment (enudler) [inline src/reports/models/database/cassandra/cassandraConnector.js]: maybe rename to joinReportsWIthSubscribers

File: src/reports/models/finalReportGenerator.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+const STATS_INTERVAL = 30;
+
+module.exports.createFinalReport = async (testId, reportId) => {
+    let stats, report;
+    report = await reportsManager.getReport(testId, reportId);
+    stats = await databaseConnector.getStats(testId, reportId);
+
+    if (stats.length === 0) {
+        let errorMessage = `Can not generate artillery report as testId: ${testId} and reportId: ${reportId} is not found`;
```
Comment (enudler) [inline src/reports/models/finalReportGenerator.js]: rename artillery -> predator

File: src/reports/models/finalReportGenerator.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    });
+
+    if (report.parallelism > 1) {
+        const bucketIntemerdiates = _.groupBy(reportInput.intermediate, 'bucket');
+        reportInput.intermediate = _.keysIn(bucketIntemerdiates).map((bucket) => {
+            return createAggregateManually(bucketIntemerdiates[bucket]);
+        });
+    }
+
+    if (_.isEmpty(reportInput.final_report)) {
```
Comment (enudler) [inline src/reports/models/finalReportGenerator.js]: remove final

File: src/reports/models/reportsManager.js [hunk 7]
```diff
@@ -1,9 +1,15 @@
...
+    uuid = require('uuid/v4');
+
 const databaseConnector = require('./databaseConnector'),
     jobConnector = require('../../jobs/models/jobManager'),
     configHandler = require('../../configManager/models/configHandler'),
-    statsConsumer = require('./statsConsumer');
+    notifier = require('./notifier'),
+    constants = require('../utils/constants');
+
+const MINIMUM_WAIT_FOR_DELAYED_REPORT_UPDATE_MS = 30000;
```
Comment (enudler) [inline src/reports/models/reportsManager.js]: dont forget move to configuration

File: src/reports/models/reportsManager.js [hunk 8]
```diff
@@ -39,32 +45,49 @@ module.exports.getLastReports = async (limit) => {
 module.exports.postReport = async (testId, reportBody) => {
     const startTime = new Date(Number(reportBody.start_time));
     const job = await jobConnector.getJob(reportBody.job_id);
+    const phase = '0';
```
Comment (enudler) [inline src/reports/models/reportsManager.js]: do we support aggreation only with 1 phase? if yes lets open an issue in the project

File: src/reports/models/reportsManager.js [hunk 9]
```diff
@@ -39,32 +45,49 @@ module.exports.getLastReports = async (limit) => {
...
+
+    await databaseConnector.updateSubscribers(testId, reportId, stats.runner_id, stats.phase_status);
+    const report = await module.exports.getReport(testId, reportId);
+    await databaseConnector.insertStats(stats.runner_id, report.test_id, report.report_id, uuid(), statsTime, report.phase, stats.phase_status, stats.data);
+    await module.exports.updateReport(report, stats);
+
     return stats;
 };
 
+module.exports.updateReport = async (report, stats) => {
```
[the remark this answers is not in the document]
Comment (NivLipetz) [reply src/reports/models/reportsManager.js]: nope

File: src/reports/models/reportsManager.js [hunk 10]
```diff
@@ -39,32 +45,49 @@ module.exports.getLastReports = async (limit) => {
...
+        reportBody.test_type, phase, startTime, reportBody.test_name,
+        reportBody.test_description, JSON.stringify(testConfiguration), job.notes, Date.now());
+    await databaseConnector.subscribeRunner(testId, reportBody.report_id, reportBody.runner_id);
     return reportBody;
 };
 
 module.exports.postStats = async (testId, reportId, stats) => {
-    await statsConsumer.handleMessage(testId, reportId, stats);
+    const statsTime = new Date(Number(stats.stats_time));
+    await notifier.notifyIfNeeded(testId, reportId, stats);
```
[the remark this answers is not in the document]
Comment (NivLipetz) [reply src/reports/models/reportsManager.js]: truth

File: src/reports/models/database/cassandra/cassandraConnector.js (hunk 4 above)
Comment (NivLipetz) [reply src/reports/models/database/cassandra/cassandraConnector.js]: done

File: src/reports/models/finalReportGenerator.js (hunk 5 above)
Comment (NivLipetz) [reply src/reports/models/finalReportGenerator.js]: done

File: src/reports/models/finalReportGenerator.js (hunk 6 above)
Comment (NivLipetz) [reply src/reports/models/finalReportGenerator.js]: done

File: src/reports/models/reportsManager.js [hunk 11]
```diff
@@ -79,20 +102,25 @@ function getReportResponse(summaryRow, config) {
         end_time: summaryRow.end_time || undefined,
         phase: summaryRow.phase,
         duration_seconds: (new Date(timeEndOrCurrent).getTime() - new Date(summaryRow.start_time).getTime()) / 1000,
-        avg_response_time_ms: lastStats.latency ? lastStats.latency.median : undefined,
+        avg_response_time_ms: undefined,
```
[the remark this answers is not in the document]
Comment (NivLipetz) [reply src/reports/models/reportsManager.js]: done

File: src/reports/models/database/cassandra/cassandraConnector.js (hunk 3 above)
Comment (NivLipetz) [reply src/reports/models/database/cassandra/cassandraConnector.js]: done

File: src/reports/models/reportsManager.js (hunk 1 above)
Comment (NivLipetz) [reply src/reports/models/reportsManager.js]: genius

File: src/reports/models/reportsManager.js (hunk 2 above)
Comment (NivLipetz) [reply src/reports/models/reportsManager.js]: done

File: src/reports/controllers/reportsController.js [hunk 12]
```diff
@@ -1,17 +1,18 @@
 
 'use strict';
-let artilleryReportGenerator = require('../models/artilleryReportGenerator');
+let finalReportGenerator = require('../models/finalReportGenerator');
 let reports = require('../models/reportsManager');
 
 module.exports.getHtmlReport = async function (req, res, next) {
```
[the remark this answers is not in the document]
Comment (NivLipetz) [reply src/reports/controllers/reportsController.js]: done - getAggregateReport

File: src/reports/models/finalReportGenerator.js [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    stats.forEach(stat => {
+        let data;
+        try {
+            data = JSON.parse(stat.data);
+            data.bucket = Math.floor((new Date(data.timestamp).getTime() - new Date(report.start_time).getTime()) / 1000 / STATS_INTERVAL) * STATS_INTERVAL;
+            switch (stat.phase_status) {
+            case 'intermediate':
+                reportInput.intermediate.push(data);
+                break;
+            case 'final_report':
```
[the remark this answers is not in the document]
Comment (NivLipetz) [reply src/reports/models/finalReportGenerator.js]: done

File: src/reports/models/finalReportGenerator.js [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+let math = require('mathjs');
+
+let logger = require('../../common/logger');
+let databaseConnector = require('./databaseConnector');
+let reportsManager = require('./reportsManager');
+let constants = require('../utils/constants');
+
+const STATS_INTERVAL = 30;
+
+module.exports.createFinalReport = async (testId, reportId) => {
```
[the remark this answers is not in the document]
Comment (NivLipetz) [reply src/reports/models/finalReportGenerator.js]: done

File: src/reports/models/database/cassandra/cassandraConnector.js [hunk 15]
```diff
@@ -75,6 +81,30 @@ function getStats(testId, reportId) {
...
 }
 
+function subscribeRunner(testId, reportId, runnerId) {
+    let params;
+    params = [testId, reportId, runnerId, constants.SUBSCRIBER_INITIALIZING_STAGE];
+    return executeQuery(SUBSCRIBE_RUNNER, params, queryOptions);
+}
+
+async function updateSubscribers(testId, reportId, runnerId, stage) {
+    const report = await getReport(testId, reportId);
```
[the remark this answers is not in the document]
Comment (NivLipetz) [reply src/reports/models/database/cassandra/cassandraConnector.js]: removed

File: src/reports/models/reportsManager.js (hunk 8 above)
Comment (NivLipetz) [reply src/reports/models/reportsManager.js]: yes, added

File: src/reports/models/reportsManager.js [hunk 16]
```diff
@@ -39,32 +45,49 @@ module.exports.getLastReports = async (limit) => {
...
     return stats;
 };
 
+module.exports.updateReport = async (report, stats) => {
+    const statsParsed = JSON.parse(stats.data);
+    const statsTime = statsParsed.timestamp;
+    const reportStatus = calculateReportStatus(report);
+    const endTime = reportStatus === constants.REPORT_FINISHED_STATUS || reportStatus === constants.REPORT_PARTIALLY_FINISHED_STATUS
+        ? statsTime : undefined;
+    await databaseConnector.updateReport(report.test_id, report.report_id, report.phase, statsTime, endTime);
```
[the remark this answers is not in the document]
Comment (NivLipetz) [reply src/reports/models/reportsManager.js]: agreed, added to issues

File: src/reports/models/reportsManager.js [hunk 17]
```diff
@@ -39,32 +45,49 @@ module.exports.getLastReports = async (limit) => {
...
     return reportBody;
 };
 
 module.exports.postStats = async (testId, reportId, stats) => {
-    await statsConsumer.handleMessage(testId, reportId, stats);
+    const statsTime = new Date(Number(stats.stats_time));
+    await notifier.notifyIfNeeded(testId, reportId, stats);
+
+    await databaseConnector.updateSubscribers(testId, reportId, stats.runner_id, stats.phase_status);
+    const report = await module.exports.getReport(testId, reportId);
```
[the remark this answers is not in the document]
Comment (NivLipetz) [reply src/reports/models/reportsManager.js]: removed, calculating the end_time dynamically now as well

File: src/reports/models/reportsManager.js (hunk 7 above)
Comment (NivLipetz) [reply src/reports/models/reportsManager.js]: removed.. no more delayed

**Pull request merged**